### PR TITLE
fix(orchestrator): cap context snapshot imports

### LINKED ISSUE #1798: Security: harden upload/import size limits against resource exhaustion
## Description
Implement an LLM-friendly, independently shippable backlog item for **harden upload/import size limits against resource exhaustion**. This issue should stay narrow enough for one agent/PR while producing durable behavior, tests, and documentation where appropriate.

## Affected locations
- Input parsing and untrusted text handling
- Web/API/control-plane boundaries
- Approval, token, and state mutation paths

## Recommendation
Add the smallest safe control that addresses this risk, prefer deny-by-default behavior for unsafe paths, document operator override semantics, and include regression tests for both allowed and denied flows.

## Goals
Deliver harden upload/import size limits against resource exhaustion as a discrete capability or hardening change.
Make the behavior deterministic enough for autonomous workers to verify.
Keep the implementation small enough for a one-issue worker and one PR.
Capture operator/user-facing behavior in docs or help text when applicable.

## Expected outcomes
harden upload/import size limits against resource exhaustion is available, documented, and test-backed.
Future PM/worker handoffs can point at this issue without needing extra interpretation.
Failure modes are explicit and produce actionable messages instead of silent drift.
The change does not broaden into adjacent backlog work or unrelated refactors.

## Suggested implementation slices
Inventory the current code/docs paths that own this behavior.
Add or update the narrow model/config/schema needed for the feature or guardrail.
Implement the smallest runtime/docs change that satisfies the acceptance criteria.
Add focused fixtures/tests for success, failure, and edge cases.
Update docs/help output and handoff notes so later agents can use it.

## Acceptance criteria
- [ ] There is a dedicated implementation or documentation change for harden upload/import size limits against resource exhaustion.
- [ ] The control defaults to the safest practical behavior and requires explicit override for risky cases.
- [ ] At least one automated test, fixture, or deterministic verifier covers the intended behavior.
- [ ] Sensitive payloads/secrets are not printed in logs, test output, issue comments, or telemetry.
- [ ] At least one negative/edge case is covered so the failure mode is explicit.
- [ ] Docs/help text or inline operator guidance explains how to use or interpret the behavior.
- [ ] The PR description links this issue and summarizes verification evidence.
- [ ] No unrelated production behavior is changed outside the stated scope.

## LLM handoff notes
- Treat this as a one-issue/one-PR unit.
- Start by searching for existing implementations and tests before adding new surfaces.
- Prefer structured outputs and fixtures over prose-only behavior.
- If this overlaps an existing open issue, comment with evidence and stop for PM deduplication instead of creating an unscoped implementation.
- Do not perform broad refactors or change unrelated worker/PM behavior while implementing this ticket.

## Source
Created by the 2026-07-11 fourth strategic backlog expansion run requesting 100 issues. Topic: **Vulnerabilities**.

### PR DESCRIPTION
**Pull request by djm204**
## Summary
- add a default 1 MiB pre-parse cap for context snapshot imports
- expose `ContextSnapshotSizeError`, `DEFAULT_CONTEXT_SNAPSHOT_MAX_BYTES`, and `LoadContextOptions` for explicit trusted overrides
- document the fail-closed snapshot import behavior

## Verification
- `TMPDIR=$PWD/.tmp npm test --workspace=@franken/orchestrator -- tests/unit/resilience/context-serializer.test.ts`
- `npm test --workspace=@franken/orchestrator -- --no-cache --pool=threads --maxWorkers=1 --no-file-parallelism tests/unit/resilience/context-serializer.test.ts`
- `npm run typecheck --workspace=@franken/orchestrator`
- `npm run build --workspace=@franken/orchestrator`

Closes #1798

### CODE REVIEW

**Comment by djm204** [comment]
@codex review

File: packages/franken-orchestrator/src/resilience/context-serializer.ts [hunk 1]
```diff
@@ -96,8 +119,21 @@ export async function saveContext(ctx: BeastContext, filePath: string): Promise<
...
+}
+
+/** Load context snapshot from a size-limited JSON file. */
+export async function loadContext(filePath: string, options: LoadContextOptions = {}): Promise<BeastContext> {
+  const maxBytes = normalizeMaxBytes(options.maxBytes);
+  const stats = await stat(filePath);
+  if (stats.size > maxBytes) {
+    throw new ContextSnapshotSizeError(filePath, stats.size, maxBytes);
+  }
   const raw = await readFile(filePath, 'utf-8');
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/resilience/context-serializer.ts]: Fixed in c5629eb8 by rejecting non-regular snapshot import paths before reading. `loadContext()` now checks `stats.isFile()` after `stat()` and throws `ContextSnapshotFileTypeError` for directories/FIFOs/procfs-style paths; added a regression test covering non-regular paths plus reran the targeted test, typecheck, and build.

**Comment by djm204** [comment]
@codex review

File: packages/franken-orchestrator/src/resilience/context-serializer.ts [hunk 2]
```diff
@@ -96,8 +126,24 @@ export async function saveContext(ctx: BeastContext, filePath: string): Promise<
...
+  if (!Number.isSafeInteger(resolved) || resolved <= 0) {
+    throw new RangeError('Context snapshot import maxBytes must be a positive safe integer');
+  }
+  return resolved;
+}
+
+/** Load context snapshot from a size-limited JSON file. */
+export async function loadContext(filePath: string, options: LoadContextOptions = {}): Promise<BeastContext> {
+  const maxBytes = normalizeMaxBytes(options.maxBytes);
+  const stats = await stat(filePath);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/resilience/context-serializer.ts]: Fixed in 75b3a2b6 by enforcing the limit on bytes actually read from the opened descriptor. `loadContext()` now opens the file first, verifies the descriptor is a regular file, rejects metadata sizes above the cap, then reads in bounded chunks and throws `ContextSnapshotSizeError` if the consumed bytes exceed `maxBytes`, covering append/replace-after-stat races. Reran targeted snapshot tests, orchestrator typecheck, and orchestrator build.

**Comment by djm204** [comment]
@codex review

File: packages/franken-orchestrator/src/resilience/context-serializer.ts [hunk 3]
```diff
@@ -96,9 +126,52 @@ export async function saveContext(ctx: BeastContext, filePath: string): Promise<
...
+/** Load context snapshot from a size-limited JSON file. */
+export async function loadContext(filePath: string, options: LoadContextOptions = {}): Promise<BeastContext> {
+  const maxBytes = normalizeMaxBytes(options.maxBytes);
+  const raw = await readRegularFileWithinLimit(filePath, maxBytes);
   const snapshot: ContextSnapshot = JSON.parse(raw);
   return deserializeContext(snapshot);
 }
+
+async function readRegularFileWithinLimit(filePath: string, maxBytes: number): Promise<string> {
+  const handle = await open(filePath, 'r');
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-orchestrator/src/resilience/context-serializer.ts]: Fixed in 6b714ead by opening context snapshots with `O_NONBLOCK` before descriptor validation, so FIFO paths do not block waiting for a writer. Added a POSIX FIFO regression test that verifies `loadContext()` rejects the FIFO immediately with `ContextSnapshotFileTypeError`; reran targeted tests, orchestrator typecheck, and orchestrator build.

**Comment by djm204** [comment]
@codex review
